jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java

```diff
@@ -0,0 +1,6131 @@
+	 * @return boolean value indicating whether the field was updated
+	 */
+	public final boolean weakCompareAndSetByte(Object obj, long offset, byte compareValue, byte setValue) {
+		return compareAndSetByte(obj, offset, compareValue, setValue);
+	}
+
+	/**
+	 * Sets the parameter value at offset in obj if the compare value 
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of getAcquire.
+	 * The set operation has the memory semantics of set.
+	 * 
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param compareValue value that is expected to be in obj at offset
+	 * @param setValue value that will be set in obj at offset if compare is successful
+	 * @return boolean value indicating whether the field was updated
+	 */
+	public final boolean weakCompareAndSetByteAcquire(Object obj, long offset, byte compareValue, byte setValue) {
+		return weakCompareAndSetByte(obj, offset, compareValue, setValue);
+	}
+
+	/**
+	 * Sets the parameter value at offset in obj if the compare value 
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of get.
+	 * The set operation has the memory semantics of setRelease.
+	 * 
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param compareValue value that is expected to be in obj at offset
+	 * @param setValue value that will be set in obj at offset if compare is successful
+	 * @return boolean value indicating whether the field was updated
+	 */
+	public final boolean weakCompareAndSetByteRelease(Object obj, long offset, byte compareValue, byte setValue) {
+		return weakCompareAndSetByte(obj, offset, compareValue, setValue);
+	}
+
+	/**
+	 * Sets the parameter value at offset in obj if the compare value 
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of get.
+	 * The set operation has the memory semantics of set.
+	 * 
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param compareValue value that is expected to be in obj at offset
+	 * @param setValue value that will be set in obj at offset if compare is successful
+	 * @return boolean value indicating whether the field was updated
+	 */
+	public final boolean weakCompareAndSetBytePlain(Object obj, long offset, byte compareValue, byte setValue) {
+		return weakCompareAndSetByte(obj, offset, compareValue, setValue);
+	}
+
+	/**
+	 * Atomically sets the parameter value at offset in obj if the compare value 
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of getAcquire.
+	 * The set operation has the memory semantics of set.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj

```

Comment:
What is this checking for? I don't understand the `>> 32`

Reply:
This question is about the implementation of checkNativeAddress()

Reply:
This is checking that  the long type address parameter does not contain a number that exceeds 32 bits if the address size is 4 bytes. I'm updating the description to reflect that.

Reply:
Isn't that what is32BitClean() is doing? I don't understand what checkNativeAddress() is doing.

Reply:
`checkNativeAddress()` passes with negative sign extended numbers while `is32BitClean()` would fail if passed a negative value because it uses logical shift right. I think this is to support sign extended pointers.

Reply:
why the `& -2L`. This will allow many numbers greater than 32-bits to pass the test.

Reply:
`shiftedValue` after line 5708 will be either -1 if the address is 32 bits and negative or 0 if the address is 32 bits and positive.

`(shiftedValue + 1) & -2L` will only equate to zero if shiftedValue is either -1 or 0.

Reply:
Ok, please add that as a comment.

Reply:
Okay. Done.